### PR TITLE
Experiment copy updates

### PR DESCRIPTION
**Pull request by kafkasl**
# Changes
### Code changes

I removed the `sandbox_id` prefix feature from the `experiment copy` because it was breaking often due to relying on objects' structure.
Also removed the function updating the gem2s hash because it is no longer needed (actually recomputing the hash had the opposite effect of forcing to run the pipeline again because the hash is no longer computed as `utils` thought).
I refactored a bit the `s3` & `dynamodb` copies because they had a weird structure of nested loops. Now, the loops are: for each experiment -> for each bucket / table.

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [x] New commands have been added to Makefile/test
- [x] Run make test
- [x] Test any modified command

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [x] Relevant Github READMEs updated
- [x] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team

### Optional
- [x] Photo of a cute animal attached to this PR

image

### CODE REVIEW

File: biomage/experiment/copy.py [hunk 1]
```diff
@@ -1,9 +1,10 @@
+import os
```
Comment (robioson) [inline biomage/experiment/copy.py:1]: Unused import

File: biomage/experiment/utils.py [hunk 2]
```diff
@@ -141,84 +141,19 @@ def get_cognito_username(email):
...
-        Key={"experimentId": experiment_id},
-        ProjectionExpression='projectId'
-    ).get("Item")['projectId']
+        Key={"experimentId": experiment_id}, ProjectionExpression="projectId"
+    ).get("Item")["projectId"]
 
     return project_id
 
 
-def create_gem2s_hash(experiment, project, samples):
```
Comment (robioson) [inline biomage/experiment/utils.py:154]: `hashlib`, `collections`, `constants` now unused imports

File: biomage/experiment/utils.py [hunk 3]
```diff
@@ -141,84 +141,19 @@ def get_cognito_username(email):
...
 def get_experiment_project_id(experiment_id, source_table):
 
-    table = boto3.resource('dynamodb').Table(source_table)
+    table = boto3.resource("dynamodb").Table(source_table)
 
     project_id = table.get_item(
-        Key={"experimentId": experiment_id},
-        ProjectionExpression='projectId'
-    ).get("Item")['projectId']
+        Key={"experimentId": experiment_id}, ProjectionExpression="projectId"
```
Comment (robioson) [inline biomage/experiment/utils.py:147]: I preferred the old formatting with each argument on a newline

File: biomage/experiment/utils.py [hunk 4]
```diff
@@ -235,4 +170,6 @@ def add_env_user_to_experiment(cfg):
 
     user_name = get_cognito_username(email=email)
 
-    return add_user_to_rbac(user_name=user_name, cfg=cfg)
+    add_user_to_rbac(user_name=user_name, cfg=cfg)
```
Comment (robioson) [inline biomage/experiment/utils.py:173]: I would rather not use named arguments here

File: biomage/utils/data.py [hunk 5]
```diff
@@ -68,20 +35,22 @@ def definitely_equal(target, source):
     return same_etag
 
 
-def copy_s3_files(sandbox_id, prefix, source_bucket, target_bucket):
+def copy_s3_bucket(prefix, source_bucket, target_bucket):
     """
-    Copy s3 files in a bucket under a prefix
+    Copy s3 files in a bucket under a prefix. Prefix is normally an experiment or
+     project id
```
Comment (robioson) [inline biomage/utils/data.py:41]: Remove leading space

File: biomage/utils/data.py [hunk 6]
```diff
@@ -108,174 +77,98 @@ def copy_s3_files(sandbox_id, prefix, source_bucket, target_bucket):
                 )
 
 
-def copy_dynamodb_records(
-    sandbox_id, staging_experiments, source_table, target_table, config
-):
+def copy_s3_data(experiments, config, origin, destination):
+    buckets = config["source-buckets"]
+    for experiment_id in experiments:
+
```
Comment (robioson) [inline biomage/utils/data.py:83]: I would remove this blank line but you can keep it if you prefer it

**Review by robioson — CHANGES_REQUESTED** [review]
Really nice code in general, I have some formatting quibbles and unused imports that should be trivial to fix.

File: README.md [hunk 7]
```diff
@@ -53,8 +53,9 @@ then simply do:
   as the first part of the name of the staging environments created by you:
   `${BIOMAGE_NICK:-${USER}}-...`.
 
-* `BIOMAGE_EMAIL` is used by the `experiment pull` command to automatically add permissions for the downloaded experiment.
-You should specify the email used to log in into the platform in staging.
+* `BIOMAGE_EMAIL` is used by the `experiment pull/copy` command to automatically add permissions for the downloaded experiment.
+You should specify the email used to log in into the platform in staging. **Note** your current aws account will need permissions
+for cognito.
```
Comment (robioson) [inline README.md:58]: Forgot to add to review, `AWS`, `Cognito`

File: biomage/experiment/utils.py (hunk 4 above)
Comment (kafkasl) [reply biomage/experiment/utils.py:173]: why not, I'd rather use them almost always

File: README.md (hunk 7 above)
Comment (kafkasl) [reply README.md:58]: what?

File: biomage/experiment/utils.py (hunk 2 above)
Comment (kafkasl) [reply biomage/experiment/utils.py:154]: removed, I don't know why but the import auto-format and remove no longer works. It seems we did fuck it up in the end

File: biomage/experiment/utils.py (hunk 3 above)
Comment (kafkasl) [reply biomage/experiment/utils.py:147]: auto-formatter sets it like that, can't change

**Comment by kafkasl** [comment]
@robioson I will be out next week, there are some comments I didn't understand, I addressed the rest. Feel free to merge (or change what you want and then merge if you want to).

I think we somehow messed up with the VS code formatting when we discussed this because the imports should be auto-deleted and they aren't.

File: README.md (hunk 7 above)
Comment (robioson) [reply README.md:58]: I mean the capitalisation is missing

File: biomage/experiment/utils.py (hunk 4 above)
Comment (robioson) [reply biomage/experiment/utils.py:170]: Because the parameters are named exactly the same as the arguments I just think it looks weird

**Review by robioson — APPROVED** [review]

**Pull request merged**